File: pattern-diagnose/assets/examples/python/strategy/before/shipping.py
# SPDX-License-Identifier: MIT
"""Before: long conditional on type — Strategy candidate."""

from dataclasses import dataclass


@dataclass
class Order:
    weight: float
    destination: str
    kind: str  # "standard" | "express" | "overnight"


def calculate_shipping(order: Order) -> float:
    if order.kind == "standard":
        return order.weight * 0.5 + 3
    elif order.kind == "express":
        return order.weight * 1.2 + 8
    elif order.kind == "overnight":
        base = order.weight * 2.5 + 15
        return base * 1.5 if order.destination.startswith("INTL-") else base
    raise ValueError(f"unknown kind: {order.kind}")

File: pattern-diagnose/assets/examples/typescript/strategy/before/shipping.ts
// SPDX-License-Identifier: MIT
// Before: long conditional on type — Strategy candidate

type ShippingKind = "standard" | "express" | "overnight";

interface Order {
  weight: number;
  destination: string;
  kind: ShippingKind;
}

export function calculateShipping(order: Order): number {
  if (order.kind === "standard") {
    return order.weight * 0.5 + 3;
  } else if (order.kind === "express") {
    return order.weight * 1.2 + 8;
  } else if (order.kind === "overnight") {
    const base = order.weight * 2.5 + 15;
    return order.destination.startsWith("INTL-") ? base * 1.5 : base;
  }
  throw new Error(`unknown kind: ${order.kind}`);
}
